### ISSUE TITLE
[LogTest testFileLoggingUsePlainText] Failure

### ISSUE THREAD

**Issue by jayahariv** [couchbase/couchbase-lite-ios#2331]
When running the unit test in a specific order, file logging is not creating the binary log files. 

It can be replicated by running the `[LogTest testFileLoggingUsePlainText]` & `[LogTest testFileLoggingReEnableLogging]` tests together.

**Comment by jayahariv** [comment]
Not much familiar with the LiteCore, but below are findings. Able to find `needsTearDown` method is returning false, in the failure case.  

When running this test case alone the `sLogDirectory` is empty, and when it's running with some other test cases, `sLogEncoder[0]` is not empty. So in both cases it's not returning from `writeEncodedLogsTo` immediately.

**Comment by jayahariv** [comment]
Got failing single test case.

```
- (void) testFileLoggingDisableLogging4AllTogether {
    [[NSFileManager defaultManager] removeItemAtPath: logFileDirectory error: nil];
    
    CBLLogFileConfiguration* config = [self logFileConfig];
    config.usePlainText = YES;
    CBLDatabase.log.file.config = config;
    CBLDatabase.log.file.level = kCBLLogLevelVerbose;
    
    NSString* inputString = [[NSUUID UUID] UUIDString];
    [self writeAllLogs: inputString];
    
    NSArray* files = [self getLogsInDirectory: config.directory properties: nil onlyInfoLogs: NO];
    Assert(files.count > 0);
    
    if (_backup) {
        CBLDatabase.log.file.level = _backup.level;
        CBLDatabase.log.file.config = _backup.config;
        _backup = nil;
    }
    
    [self backupFileLogger];
    [[NSFileManager defaultManager] removeItemAtPath: logFileDirectory error: nil];
    [NSThread sleepForTimeInterval: 1.0];
    
    config = [self logFileConfig];
    config.usePlainText = YES;
    CBLDatabase.log.file.config = config;
    CBLDatabase.log.file.level = kCBLLogLevelVerbose;
    
    inputString = [[NSUUID UUID] UUIDString];
    [self writeAllLogs: inputString];
    
    files = [self getLogsInDirectory: config.directory properties: nil onlyInfoLogs: NO];
    Assert(files.count > 0);
}
```

**Comment by jayahariv** [comment]
When the contents are removed at the file URL, write after that is not creating the binary logs. `removeItemAtPath ` is returning the error below.

```
((error) == nil) failed: "Error Domain=NSCocoaErrorDomain Code=4 "“LogTestLogs” couldn’t be removed." UserInfo={NSFilePath=/var/folders/k1/80rmvdfd1_77xmz6x3n5d3zw0000gp/T/LogTestLogs, NSUserStringVariant=(
    Remove
), NSUnderlyingError=0x60400007e810 {Error Domain=NSPOSIXErrorDomain Code=2 "No such file or directory"}}"
```

**Comment by jayahariv** [comment]
This should be the case which is happening, removing the log files, while the logging happens.

```
- (void) testFileLoggingRemoveDirectoryWhileLogging {
    CBLLogFileConfiguration* config = [self logFileConfig];
    CBLDatabase.log.file.config = config;
    CBLDatabase.log.file.level = kCBLLogLevelVerbose;
    
    NSString* inputString = [[NSUUID UUID] UUIDString];
    CBLLogInfo(Database, @"%@", inputString);
    
    NSArray* files = [self getLogsInDirectory: config.directory properties: nil onlyInfoLogs: NO];
    Assert(files.count > 0);
    
    if ([[NSFileManager defaultManager] fileExistsAtPath:logFileDirectory]) {
        NSError* error;
        [[NSFileManager defaultManager] removeItemAtPath: logFileDirectory error: &error];
        AssertNil(error);
    }
    
    inputString = [[NSUUID UUID] UUIDString];
    CBLLogInfo(Database, @"%@", inputString);
    
    files = [self getLogsInDirectory: config.directory properties: nil onlyInfoLogs: NO];
    Assert(files.count > 0); // failure zero files
}
```

**Comment by pasin** [comment]
I'm confused now what the failure is. The log directory doesn't get created? If that is the case, this [code](https://github.com/couchbase/couchbase-lite-ios/blob/master/Objective-C/CBLFileLogger.mm#L102) creates the directory if it doesn't exist.

**Comment by jayahariv** [comment]
Directory is getting created but the binary files are not re-created. I guess, the `sFileOut` are using the old values, and keep writing it to output file stream. wonder whether that should raise an error is the output file stream, if file is missing. If its encoded, it's using the existing `sLogEncoders`. It is what the rough analysis found.

**Comment by pasin** [comment]
@borrrden Do you know what could cause the log files not getting created after deleting the log directory? OR Maybe this is something shouldn't be doing?

@jayahariv What if setting the log level to NONE before deleting the directory?

**Comment by borrrden** [comment]
Why would you delete the log directory while using it?  I guess it is something that is possible but there is no logic to account for this in the logger.  It simply has a file stream object that it writes to.  There is no logic in there to recreate the file if it is removed from under it.  That would mean that at every log statement we'd have to query the filesystem for the file.

**Comment by borrrden** [comment]
If you want it to always create new files for every call to `c4log_writeToBinaryDirectory` though, that's easy enough because the logic to protect against log rotation (`needsTeardown`) is obsolete now that we have CBL config as an atomic operation.

**Comment by jayahariv** [comment]
@borrrden I could give more context, this happened in the unit tests, CBLTestClass `setup` method was removing the specific folder. So whenever the same configuration tests are running back to back, later one will fail saying the logs files are missing. So investigated, and found this happens when folder removal happens, in between two unit tests with the same configuration.

**Comment by borrrden** [comment]
Alright after extensive discussion between many people here is the summary:

This is a unix thing.  The file is *not* deleted despite it not existing where the test thinks that it does.  It won't be "officially" deleted until the logging stream is closed.  

Because of this, the log rotation logic should still be in effect and eventually a new log will be created once max size is reached.  That makes this issue not so severe in my opinion.  If the user is going to be deleting logs from the directory that is currently in use for logging then the consequence will be they don't have logs for whatever period of time between then, and either the rotation or app restart.

For my logging tests, I noticed that reusing the same log directory for every test was causing other problems as well (e.g. files leftover from a previous run would be present in the next, etc) so in the end I used a different directory for each test.
